File: ArchiveProcessor/ArchiveCapture/app/src/main/java/com/archiveprocessor/capture/net/MacEndpoint.kt
package com.archiveprocessor.capture.net

import org.json.JSONObject

/**
 * Connection info for the Mac receiver.
 *
 * - **LAN/USB** (`mode = "lan"`): decoded from the pairing QR `{host, port, token, name}`; the phone talks
 *   HTTP to `baseUrl`. `token` is the Bearer pairing code.
 * - **Cloud** (`mode = "cloud"`): the phone chose Cloud. `host`/`port` are unused; the relay token comes
 *   from [relayToken]. The Mac now emits ONE combined LAN QR that ALSO carries the relay token in an
 *   OPTIONAL `relay` key (see below), so cloud is reached from that same scan — a legacy `{mode:"cloud",…}`
 *   QR is still accepted for back-compat. The relay token is the same value the Mac writes as the Drive
 *   folder's `appProperties.relayToken` and into `_epoch.json`, which `DriveRelayTransport` uses to
 *   self-discover the shared folder. `account` is an optional Google-account email hint for sign-in.
 *
 * `relay` is the OPTIONAL combined-QR field: present on a current Mac's QR when it's signed into Drive,
 * absent on an older / LAN-only QR (→ null, LAN pairing unaffected).
 */
data class MacEndpoint(
    val host: String,
    val port: Int,
    val token: String,
    val name: String,
    val mode: String = "lan",
    val account: String? = null,
    val relay: String? = null
) {
    val baseUrl: String get() = "http://$host:$port"
    val isCloud: Boolean get() = mode == "cloud"

    /** The Drive-relay token to use for the Cloud transport: the explicit combined-QR `relay` field, else
     *  the token itself when this endpoint was already built as a cloud endpoint. Null for a plain LAN QR
     *  with no `relay` field (→ Cloud isn't available from that scan; sign the Mac into Drive first). */
    val relayToken: String? get() = relay ?: token.takeIf { isCloud }

    companion object {
        fun fromQrPayload(payload: String): MacEndpoint? {
            return try {
                val o = JSONObject(payload)
                val token = o.optString("token", "")
                if (token.isBlank()) return null
                val name = o.optString("name", "Mac")
                val relay = o.optString("relay", "").ifBlank { null }   // OPTIONAL; tolerate absence
                if (o.optString("mode", "lan") == "cloud") {
                    MacEndpoint("", 0, token, name, "cloud", o.optString("account", "").ifBlank { null }, relay)
                } else {
                    val host = o.optString("host", "")
                    val port = o.optInt("port", 0)
                    if (host.isBlank() || port <= 0) null
                    else MacEndpoint(host, port, token, name, "lan", null, relay)
                }
            } catch (e: Exception) {
                null
            }
        }
    }
}

File: ArchiveProcessor/ArchiveCaptureiOS/Sources/ArchiveCaptureiOS/Net/MacEndpoint.swift
import Foundation

/// Connection info for the Mac receiver, decoded from the pairing QR JSON: `{host, port, token, name}`
/// PLUS an OPTIONAL `relay` token. Mirrors the Android `MacEndpoint` so the same QR works for both
/// companions.
struct MacEndpoint: Codable, Equatable {
    let host: String
    let port: Int
    let token: String
    let name: String
    /// OPTIONAL Drive-relay token from the combined pairing QR (`relay` key). Absent on an older Mac's
    /// LAN-only QR → nil (LAN pairing is unaffected). When present it carries the token the Drive cloud
    /// transport uses to find the Mac's shared folder; today it equals `token`. Optional so decoding a
    /// previously-persisted endpoint (no `relay`) still succeeds.
    var relay: String? = nil

    var baseURL: String { "http://\(host):\(port)" }

    static func fromQRPayload(_ payload: String) -> MacEndpoint? {
        guard let data = payload.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let host = obj["host"] as? String, !host.isEmpty,
              let token = obj["token"] as? String, !token.isEmpty else { return nil }
        let port: Int
        if let p = obj["port"] as? Int { port = p }
        else if let s = obj["port"] as? String, let p = Int(s) { port = p }
        else { return nil }
        guard port > 0 else { return nil }
        let relay = (obj["relay"] as? String).flatMap { $0.isEmpty ? nil : $0 }   // tolerate absence
        return MacEndpoint(host: host, port: port, token: token,
                           name: (obj["name"] as? String) ?? "Mac", relay: relay)
    }
}
